x/cdp/keeper/cdp.go

```diff
@@ -0,0 +1,407 @@
+}
+
+// GetAllCdpsByDenomAndRatio returns all cdps of a particular collateral type and below a certain collateralization ratio
+func (k Keeper) GetAllCdpsByDenomAndRatio(ctx sdk.Context, denom string, targetRatio sdk.Dec) (cdps types.CDPs) {
+	k.IterateCdpsByCollateralRatio(ctx, denom, targetRatio, func(cdp types.CDP) bool {
+		cdps = append(cdps, cdp)
+		return false
+	})
+	return
+}
+
+// SetNextCdpID sets the highest cdp id in the store
+func (k Keeper) SetNextCdpID(ctx sdk.Context, id uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CdpIDKey)
+	store.Set([]byte{}, types.GetCdpIDBytes(id))
+}
+
+// GetNextCdpID returns the highest cdp id from the store
+func (k Keeper) GetNextCdpID(ctx sdk.Context) (id uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CdpIDKey)
+	bz := store.Get([]byte{})
+	if bz == nil {
+		panic("starting cdp id not set in genesis")
+	}
+	id = types.GetCdpIDFromBytes(bz)
+	return
+}
+
+// IndexCdpByOwner sets the cdp id in the store, indexed by the owner
+func (k Keeper) IndexCdpByOwner(ctx sdk.Context, cdp types.CDP) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CdpIDKeyPrefix)
+	cdpIDs, found := k.GetCdpIdsByOwner(ctx, cdp.Owner)
+
+	if !found {
+		idBytes := k.cdc.MustMarshalBinaryLengthPrefixed([]uint64{cdp.ID})
+		store.Set(cdp.Owner, idBytes)
+		return
+	}
+	for _, id := range cdpIDs {
+		if id == cdp.ID {
+			return
+		}
+		cdpIDs = append(cdpIDs, cdp.ID)
+		store.Set(cdp.Owner, k.cdc.MustMarshalBinaryLengthPrefixed(cdpIDs))
+	}
+}
+
+// RemoveCdpOwnerIndex deletes the cdp id from the store's index of cdps by owner
+func (k Keeper) RemoveCdpOwnerIndex(ctx sdk.Context, cdp types.CDP) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CdpIDKeyPrefix)
+	cdpIDs, found := k.GetCdpIdsByOwner(ctx, cdp.Owner)
+	if !found {
+		return
+	}
+	updatedCdpIds := []uint64{}
+	for _, id := range cdpIDs {
+		if id != cdp.ID {
+			updatedCdpIds = append(updatedCdpIds, id)
+		}
+	}
+	if len(updatedCdpIds) == 0 {
+		store.Delete(cdp.Owner)
+	}
+	store.Set(cdp.Owner, k.cdc.MustMarshalBinaryLengthPrefixed(updatedCdpIds))
+
+}
+
+// IndexCdpByCollateralRatio sets the cdp id in the store, indexed by the collateral type and collateral to debt ratio
+func (k Keeper) IndexCdpByCollateralRatio(ctx sdk.Context, denom string, id uint64, collateralRatio sdk.Dec) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CollateralRatioIndexPrefix)
+	db, _ := k.GetDenomPrefix(ctx, denom)
+	store.Set(types.CollateralRatioKey(db, id, collateralRatio), types.GetCdpIDBytes(id))
+}
+
+// RemoveCdpCollateralRatioIndex deletes the cdp id from the store's index of cdps by collateral type and collateral to debt ratio
+func (k Keeper) RemoveCdpCollateralRatioIndex(ctx sdk.Context, denom string, id uint64, collateralRatio sdk.Dec) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CollateralRatioIndexPrefix)
+	db, _ := k.GetDenomPrefix(ctx, denom)
+	store.Delete(types.CollateralRatioKey(db, id, collateralRatio))
+}
```

Comment:
In the `staking` module, any index updates are written into the Get/Set methods. I kind of like that approach - reduces code repitition in higher level keeper methods, and avoids the posibility that indexes are not updated when they should be.
Do you think it's worth doing for CDP indexes?

Reply:
Not quite sure what the suggestion is here. Looking through `staking`, the `Set` methods are pretty small and there are separate methods for indexes (ie `SetValidatorByPowerIndex`). The main difference appears to be that there is only one key prefix, which I'm not sure I can accomplish. I thought about the various indexes and iterators that were desirable for CDPs and couldn't come up with an approach that had fewer keys.

Reply:
ah yeah looking at staking again it's only the `SetUnbondingDelegations` and `SetRedelegations` methods that update their indexes. Didn't see validator indexes are not updated in `SetValidator`.

The idea was that I couldn't think of any reason you would want to set a CDP in the store, but not update the CDP indexes. Having this happen by accident could be bad. Therefore it seems safer to combine SetCDP and IndexCdpByCollateralRatio into one function.